src/meta/src/hummock/compaction/tier_compaction_picker.rs

```diff
@@ -39,145 +39,78 @@ impl TierCompactionPicker {
+            if level.level_type == non_overlapping_type
+                && level.total_file_size > self.config.sub_level_max_compaction_bytes
             {
                 continue;
             }
 
-            if level
-                .table_infos
-                .iter()
-                .any(|table| level_handler.is_pending_compact(&table.id))
-            {
+            if level_handler.is_level_pending_compact(level) {
                 continue;
             }
 
-            let mut compaction_bytes = level.total_file_size;
             let mut select_level_inputs = vec![InputLevel {
                 level_idx: 0,
                 level_type: level.level_type,
                 table_infos: level.table_infos.clone(),
             }];
+
             let max_compaction_bytes = std::cmp::min(
                 self.config.max_compaction_bytes,
-                self.config.max_bytes_for_level_base,
+                self.config.max_bytes_for_level_base * 2,
             );
 
+            let mut compaction_bytes = level.total_file_size;
+            let mut max_level_size = level.total_file_size;
+            let mut compact_file_count = level.table_infos.len();
+
             for other in &l0.sub_levels[idx + 1..] {
                 if compaction_bytes >= max_compaction_bytes {
                     break;
                 }
 
-                if level.total_file_size > self.config.sub_level_max_compaction_bytes {
+                if level.level_type == non_overlapping_type
+                    && level.total_file_size > self.config.sub_level_max_compaction_bytes
+                {
                     break;
                 }
 
-                if other
-                    .table_infos
-                    .iter()
-                    .any(|table| level_handler.is_pending_compact(&table.id))
-                {
+                if level_handler.is_level_pending_compact(other) {
                     break;
                 }
 
                 compaction_bytes += other.total_file_size;
+                compact_file_count += other.table_infos.len();
+                max_level_size = std::cmp::max(max_level_size, other.total_file_size);
                 select_level_inputs.push(InputLevel {
                     level_idx: 0,
                     level_type: other.level_type,
                     table_infos: other.table_infos.clone(),
                 });
             }
 
-            if select_level_inputs
-                .iter()
-                .map(|level| level.table_infos.len())
-                .sum::<usize>()
-                < self.config.level0_tier_compact_file_number as usize
-            {
+            if compact_file_count < self.config.level0_tier_compact_file_number as usize {
                 continue;
             }
 
-            if select_level_inputs.len() <= self.config.level0_tier_compact_file_number as usize / 2
-            {
+            let is_write_amp_large =
```

Comment:
Could you add some comments to describe how to calculate is_write_amp_large and why the l0 file_number need to divide 2 ?

Reply:
OK.